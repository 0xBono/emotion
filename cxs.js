module.exports = require('./src/cxs/babel')
