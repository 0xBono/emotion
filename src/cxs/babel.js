module.exports = function (babel) {
  const {types: t} = babel

  function createClassNameAttr (expression) {
    return t.jSXAttribute(
      t.jSXIdentifier('className'),
      t.JSXExpressionContainer(expression)
    )
  }

  return {
    name: 'emotion-for-cxs', // not required
    inherits: require('babel-plugin-syntax-jsx'),
    visitor: {
      JSXOpeningElement (path, state) {
        let cssPath
        let classNamesPath

        path.get('attributes').forEach(openElPath => {
          const attrPath = openElPath.get('name')
          const name = attrPath.node.name

          if (name === 'css') {
            cssPath = attrPath
          }

          if (name === 'className') {
            classNamesPath = attrPath
          }
        })

        if (!cssPath) return

        let cssPropValue = cssPath.container && cssPath.container.value
        let classNamesValue =
          classNamesPath &&
          classNamesPath.container &&
          classNamesPath.container.value

        if (t.isJSXExpressionContainer(cssPropValue)) {
          cssPropValue = cssPropValue.expression
        }

        let cxsCssFunction = t.callExpression(t.identifier('cxs'), [
          cssPropValue
        ])

        if (!classNamesValue) {
          cssPath.parentPath.replaceWith(createClassNameAttr(cxsCssFunction))
          return
        }

        cssPath.parentPath.remove()
        if (t.isJSXExpressionContainer(classNamesValue)) {
          classNamesPath.parentPath.replaceWith(
            createClassNameAttr(
              t.binaryExpression(
                '+',
                t.binaryExpression(
                  '+',
                  classNamesValue.expression,
                  t.stringLiteral(' ')
                ),
                cxsCssFunction
              )
            )
          )
        } else {
          classNamesPath.parentPath.replaceWith(
            createClassNameAttr(
              t.binaryExpression(
                '+',
                t.binaryExpression(
                  '+',
                  t.stringLiteral(classNamesValue.value || ''),
                  t.stringLiteral(' ')
                ),
                cxsCssFunction
              )
            )
          )
        }
      }
    }
  }
}
