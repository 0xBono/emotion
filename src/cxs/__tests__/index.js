/* eslint-disable jsx-quotes,no-useless-escape */
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import plugin from '../babel'
import {matcher, serializer} from 'jest-glamor-react'
import cxs from 'cxs'

expect.addSnapshotSerializer(serializer)
expect.extend(matcher)

const babel = require('babel-core')

describe('emotion/cxs', () => {
  describe('babel', () => {
    test('basic', () => {
      const basic = `(<div className="a" css={{ color: 'brown' }}></div>)`
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('no css attr', () => {
      const basic = '(<div></div>)'
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('css empty', () => {
      const basic = '(<div css={{}}></div>)'
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('wrong value type', () => {
      const basic = '(<div css={5}></div>)'
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('no className', () => {
      const basic = `(<div css={{ color: 'brown' }}></div>)`
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('emptyClassName', () => {
      const basic = `(<div className="" css={{ color: 'brown' }}></div>)`
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('className as expression', () => {
      const basic = `(<div className={props.className} css={{ color: 'brown' }}></div>)`
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })

    test('className as expression string', () => {
      const basic =
        "(<div className={`test__class\`} css={{ color: 'brown'}} this={`hello\`}></div>)"
      const {code} = babel.transform(basic, {plugins: [plugin]})
      expect(code).toMatchSnapshot()
    })
  })

  describe('real', () => {
    test('basic', (done) => {
      const tree = renderer
        .create(
          <p css={{color: 'red'}}>
            hello world
          </p>
        )
        .toJSON()

      expect(tree).toMatchSnapshotWithGlamor()
      setTimeout(done)
    })

    test('kitchen sink', (done) => {
      const tree = renderer
        .create(
          <div className="css__legacy-stuff" css={{color: 'blue'}}>
            <h1 css={{'@media(min-width: 420px)': {fontSize: 48}}}>
              BOOM
            </h1>
            <p className="test_class1" css={{color: 'gray'}}>Hello</p>
            <p
              className="test_class1 test___class45"
              css={{border: '1px solid blue'}}
            >
              World
            </p>
            <p css={{display: 'flex'}}>
              hello world
            </p>

          </div>
        )
        .toJSON()

      expect(tree).toMatchSnapshotWithGlamor()
      setTimeout(done)
    })
  })
})
