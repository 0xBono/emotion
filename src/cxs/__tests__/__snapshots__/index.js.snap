// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emotion/cxs babel basic 1`] = `"<div className={\\"a\\" + \\" \\" + cxs({ color: 'brown' })}></div>;"`;

exports[`emotion/cxs babel className as expression 1`] = `"<div className={props.className + ' ' + cxs({ color: 'brown' })}></div>;"`;

exports[`emotion/cxs babel className as expression string 1`] = `"<div className={\`test__class\` + ' ' + cxs({ color: 'brown' })} this={\`hello\`}></div>;"`;

exports[`emotion/cxs babel css empty 1`] = `"<div className={cxs({})}></div>;"`;

exports[`emotion/cxs babel emptyClassName 1`] = `"<div className={\\"\\" + \\" \\" + cxs({ color: 'brown' })}></div>;"`;

exports[`emotion/cxs babel no className 1`] = `"<div className={cxs({ color: 'brown' })}></div>;"`;

exports[`emotion/cxs babel no css attr 1`] = `"<div></div>;"`;

exports[`emotion/cxs babel wrong value type 1`] = `"<div className={cxs(5)}></div>;"`;

exports[`emotion/cxs real basic 1`] = `
<p
  className="c-red"
>
  hello world
</p>
`;

exports[`emotion/cxs real kitchen sink 1`] = `
<div
  className="css__legacy-stuff c-blue"
>
  <h1
    className="_1eomoo"
  >
    BOOM
  </h1>
  <p
    className="test_class1 c-gray"
  >
    Hello
  </p>
  <p
    className="test_class1 test___class45 border-1px_solid_blue"
  >
    World
  </p>
  <p
    className="d-flex"
  >
    hello world
  </p>
</div>
`;
